yt/utilities/linear_interpolators.py

```diff
@@ -1,11 +1,82 @@
+import abc
+
 import numpy as np
 
 import yt.utilities.lib.interpolators as lib
 from yt.funcs import mylog
 
 
-class UnilinearFieldInterpolator:
-    def __init__(self, table, boundaries, field_names, truncate=False):
+class _LinearInterpolator(abc.ABC):
+    _ndim: int
+    _dim_i_type = "int32"
+
+    def __init__(self, table, field_names, truncate=False, *, store_table=True):
+        if store_table:
+            self.table = table.astype("float64")
+        else:
+            self.table = None
+        self.table_shape = table.shape
+        self.truncate = truncate
+        self._field_names = field_names
+
+    def _raise_truncation_error(self, data_object):
+        mylog.error(
+            "Sorry, but your values are outside "
+            "the table!  Dunno what to do, so dying."
+        )
+        mylog.error("Error was in: %s", data_object)
+        raise ValueError
+
+    def _validate_table(self, table_override):
+        if table_override is None:
+            if self.table is None:
+                msg = (
+                    f"You must either store the table used when initializing a new "
+                    f"{type(self).__name__} (set `store_table=True`) or you must provide a `table` when "
+                    f"calling {type(self).__name__}"
+                )
+                raise ValueError(msg)
+            return self.table
+
+        if table_override.shape != self.table_shape:
+            msg = f"The table_override shape, {table_override.shape}, must match the base table shape, {self.table_shape}"
+            raise ValueError(msg)
+
+        return table_override.astype("float64")
+
+    def _get_digitized_arrays(self, data_object):
+        return_arrays = []
+        for dim in "xyzw"[: self._ndim]:
+            dim_name = getattr(self, f"{dim}_name")
+            dim_bins = getattr(self, f"{dim}_bins")
+
+            dim_vals = data_object[dim_name].ravel().astype("float64")
```

Comment:
`.ravel().astype("float64")` copies data at least once, and up to two times. Inverting order of operations will make the number of copies exactly one in all cases.
```suggestion
            dim_vals = data_object[dim_name].astype("float64").ravel()
```